### PR TITLE
Add C API Guidelines document

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**: 
Add C_API _Guidelines.md so everyone has a clear guidance when adding new API

Signed-off-by: Dmitri Smirnov <dmitrism@microsoft.com>

### CODE REVIEW

File: docs/C_API_Guidelines.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+All APIs must have a proper documentation header that includes:
+
+* API summary includes any limitations, such as types that it operates on.
+* Description of each of its arguments and whether it is an in, out or in/out argument. Please, document if the user is responsible for memory deallocation or object destruction and how that can be done. Document that strings are UTF-8 encoded.
+* Describe its return value.
+
+Microsoft recommends XML format. Such a format is automatically suggested by Visual Studio when one types 3 consecutive slashes. Both C++ and C# compilers can generate XML documentation with /doc switch as described here. The documentation then can be converted to HTML pages using tools such as Sandcastle.
+
+However, XML format does not seem to be supported in C. Current style looks like Doxygen but the amount of documentation is so small and the amount of effort to document it is big so we might as well change the style of it if we have other suggestions. Otherwise, we need to try and properly document all of it.
```
Comment (pranavsharma) [inline docs/C_API_Guidelines.md:21]: Let's make a decision and document it here. This sentence seems inconclusive.

File: docs/C_API_Guidelines.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# ORT API Guidelines
+
+## Introduction
+
+Our public C API is the main interface of ONNX runtime with our customers. This document endeavors to describe our expectations with respect to the documentation and the quality of the API entry points so we are better equipped to code review new contributions and address existing shortcomings.
+
+The document is entitled guidelines. However, the expectation is that everyone understands it and adheres to it when implementing a new API or while reviewing contributions from others.
```
Comment (pranavsharma) [inline docs/C_API_Guidelines.md:7]: Is there a typo in this sentence? "The document is entitled guidelines."

File: docs/C_API_Guidelines.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+* The user may want to supply their own allocator to use. Many of our APIs do that.
+
+* The API does not have to declare a separate entry point for deallocating memory specifically for its type of allocation as we already have such entry points. OrtAllocatorAlloc/OrtAllocatorFree.
+
+Consider established patterns when APIs return multiple allocations.
+
+### 7. Public API must document that all strings they accept, and return are UTF-8 encoded
+
+All API return and accept strings in UTF-8 encodings. We must be mindful of that when maintaining language bindings.
```
Comment (pranavsharma) [inline docs/C_API_Guidelines.md:75]:
```suggestion
All APIs must return and accept strings in UTF-8 encodings. We must be mindful of that when maintaining language bindings.
```

File: docs/C_API_Guidelines.md (hunk 1 above)
Comment (yuslepukhin) [reply docs/C_API_Guidelines.md:21]: Studio puts XML comments into C# automatically. Let's mandate those there. We can make Doxygen a standard for C/C++ API although internally we can painlessly use XML.

---
In reply to: [517065928](https://github.com/microsoft/onnxruntime/pull/5686#discussion_r517065928) [](ancestors = 517065928)

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — DISMISSED** [review]

**Review by pranavsharma — APPROVED** [review]

File: docs/C_API_Guidelines.md (hunk 3 above)
Comment (Craigacp) [reply docs/C_API_Guidelines.md]: The CreateSession API still uses wchar on Windows for the path right? It's UTF-8 on macOS and Linux.

**Pull request merged**